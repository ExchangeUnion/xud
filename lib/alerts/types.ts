import { AlertType, ChannelSide } from '../constants/enums';

type BaseAlert = {
  type: AlertType;
  message: string;
  date: number;
};

export type BalanceAlertEvent = {
  /** The total balance of the channel when the alert is triggered. */
  totalBalance: number;
  /** The side of the balance either local or remote. */
  side: ChannelSide;
  /** The balance that triggered the alert. */
  sideBalance: number;
  /** The alert threshold in percentage, e.g. 10 means %10. */
  bound: number;
  /** The currency of the channel. */
  currency: string;
};
export type BalanceAlert = BaseAlert & BalanceAlertEvent;

export type Alert = BalanceAlert;
