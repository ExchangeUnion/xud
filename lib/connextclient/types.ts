/**
 * The configurable options for the connext client.
 */
export type ConnextClientConfig = {
  disable: boolean;
  host: string;
  port: number;
  webhookhost: string;
  webhookport: number;
};

/** General information about the state of this connext client. */
export type ConnextInfo = {
  status: string;
  error?: string;
  address?: string;
  chain?: string;
  version?: string;
};

/** Response for eth_blockNumber call  */
export type ConnextBlockNumberResponse = {
  result: string;
};

/** Response for eth_gasPrice call  */
export type EthproviderGasPriceResponse = {
  result: string;
};

type Balance = {
  amount: string[];
  to: string[];
};

/** Connext channel details */
export type ConnextChannelDetails = {
  assetIds: string[];
  balances: Balance[];
  channelAddress: string;
  merkleRoot: string;
  processedDepositsA: string[];
  processedDepositsB: string[];
  networkContext: {
    chainId: number;
    channelFactoryAddress: string;
    transferRegistryAddress: string;
    providerUrl: string;
  };
  nonce: number;
  alice: string;
  aliceIdentifier: string;
  bob: string;
  bobIdentifier: string;
  timeout: string;
  latestUpdate: {
    assetId: string;
    balance: Balance;
    channelAddress: string;
    details: {
      totalDepositsAlice: string;
      totalDepositsBob: string;
    };
    fromIdentifier: string;
    nonce: number;
    aliceSignature: string;
    bobSignature: string;
    toIdentifier: string;
    type: string;
  };
  defundNonce: string;
  inDispute: boolean;
};

/**
 * The payload for tokenPayment call.
 */
export type ConnextTransferRequest = {
  type: 'HashlockTransfer';
  channelAddress: string;
  amount: string;
  assetId: string;
  details: {
    lockHash: string;
    expiry: string;
  };
  recipient: string;
  meta: {
    routingId: string;
  };
  publicIdentifier: string;
};

/**
 * The response for errors.
 */
export type ConnextErrorResponse = { message: string };

export type ConnextConfig = {
  publicIdentifier: string;
  signerAddress: string;
  index: number;
};

/**
 * The response for /config call.
 */
export type ConnextConfigResponse = ConnextConfig[];

/**
 * The response for /channel call.
 */
export type ConnextChannelResponse = string[];

/**
 * The response for balance call.
 */
export type ConnextChannelBalanceResponse = {
  freeBalanceOffChain: string;
  nodeFreeBalanceOffChain: string;
  freeBalanceOnChain: string;
};

export type GetBlockByNumberResponse = {
  result: {
    difficulty: string;
    extraData: string;
    gasLimit: string;
    gasUsed: string;
    hash: string;
    logsBloom: string;
    miner: string;
    mixHash: string;
    nonce: string;
    number: string;
    parentHash: string;
    receiptsRoot: string;
    sha3Uncles: string;
    size: string;
    stateRoot: string;
    timestamp: string;
    totalDifficulty: string;
    transactions: string[];
    transactionsRoot: string;
    uncles: string[];
  };
};

/**
 * The response for hashLockTransfer call.
 */
export type ConnextTransferResponse = {
  channelAddress: string;
  transferId: string;
  routingId: string;
};

/**
 * The response for withdraw call.
 */
export type ConnextWithdrawResponse = {
  channelAddress: '0xa929dB0530daB525596f5d48Fb5C322fDa8A337B';
  transferId: '0xc2e4592d3fb6c02ee1d3b07bed83b5914f8ca084b0f91d6b80bf8107e58c9c38';
};

type ConnextRoutingPath = {
  recipient: string;
  recipientChainId: number;
  recipientAssetId: string;
};

export type ConnextTransfer = {
  channelFactoryAddress: string;
  assetId: string;
  chainId: number;
  channelAddress: string;
  balance: {
    amount: string[];
    to: string[];
  };
  initiator: string;
  responder: string;
  initialStateHash: string;
  transferDefinition: string;
  transferEncodings: string[];
  transferId: string;
  transferState: {
    lockHash: string;
    expiry: string;
  };
  transferTimeout: string;
  meta: {
    requireOnline: boolean;
    routingId: string;
    path: ConnextRoutingPath[];
    senderIdentifier?: string;
  };
  transferResolver?: {
    preImage?: string;
  };
};

export type TransfersByRoutingIdResponse = ConnextTransfer[];

export type ExpectedIncomingTransfer = {
  rHash: string;
  units: bigint;
  expiry: number;
  tokenAddress: string;
  routingId: string;
};

export type ConnextPreimageRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: string;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      requireOnline: boolean;
      routingId: string;
      path: [
        {
          recipient: string;
          recipientChainId: number;
          recipientAssetId: string;
        },
      ];
    };
    transferResolver: {
      preImage: string;
    };
  };
  conditionType: string;
};

export type ConnextIncomingTransferRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: number;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      routingId: string;
    };
  };
  conditionType: string;
};

export type ConnextDepositConfirmedRequest = {
  id: string;
  data?: {
    amount: {
      _hex: string;
    };
    assetId: string;
    hash: string;
  };
};

export type ProvidePreimageEvent = {
  rHash: string;
  preimage: string;
};

export type TransferReceivedEvent = {
  tokenAddress: string;
  rHash: string;
  expiry: number;
  units: bigint;
  routingId: string;
};
