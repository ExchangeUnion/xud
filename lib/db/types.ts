import { BelongsToGetAssociationMixin, Model } from 'sequelize';
import { ReputationEvent, SwapClientType } from '../constants/enums';
import { Currency, Order, Pair } from '../orderbook/types';
import { Address, NodeConnectionInfo } from '../p2p/types';
import { SwapDeal } from '../swaps/types';

/*
 * The following definitions are in sets of triplets, one for each Model (which represents a table in the database).
 *
 * "xCreationAttributes" is the type definition for the object which is required when a new record is to be created.
 *
 * "xAttributes" is the type definition of the record.
 *
 * "xInstance" is the type definition of a fetched record as a Sequelize row instance, which contains some util properties.
 */

/* Currency */
export type CurrencyCreationAttributes = Currency;

export type CurrencyAttributes = CurrencyCreationAttributes;

export interface CurrencyInstance extends Model<CurrencyAttributes, CurrencyCreationAttributes>, CurrencyAttributes {}

/* SwapDeal */
export type SwapDealCreationAttributes = Pick<
  SwapDeal,
  Exclude<keyof SwapDeal, 'takerMaxTimeLock' | 'price' | 'pairId' | 'isBuy' | 'takerUnits' | 'makerUnits'>
>;

export type SwapDealAttributes = SwapDealCreationAttributes & {
  /** The internal db node id of the counterparty peer for this swap deal. */
  nodeId: number;
};

export interface SwapDealInstance extends Model<SwapDealAttributes, SwapDealCreationAttributes>, SwapDealAttributes {
  getNode: BelongsToGetAssociationMixin<NodeInstance>;
  getOrder: BelongsToGetAssociationMixin<OrderInstance>;
  Node?: NodeInstance;
  Order?: OrderInstance;
}

export type OrderCreationAttributes = Pick<Order, Exclude<keyof Order, 'quantity' | 'hold' | 'price'>> & {
  /** The internal db node id of the peer that created this order. */
  nodeId?: number;
  localId?: string;
  /** The price for the order expressed in units of the quote currency. */
  price?: number;
};

export type OrderAttributes = OrderCreationAttributes & {};

export interface OrderInstance extends Model<OrderAttributes, OrderCreationAttributes>, OrderAttributes {}

export type TradeCreationAttributes = {
  /** The order id of the maker order involved in this trade. */
  makerOrderId: string;
  /** The order id of the taker order involved in this trade, if applicable. */
  takerOrderId?: string;
  /** The rHash of the swap that filled this trade, if applicable. */
  rHash?: string;
  /** The quantity transacted in this trade. */
  quantity: number;
};

export type TradeAttributes = TradeCreationAttributes & {};

export interface TradeInstance extends Model<TradeAttributes, TradeCreationAttributes>, TradeAttributes {
  getMakerOrder: BelongsToGetAssociationMixin<OrderInstance>;
  getTakerOrder: BelongsToGetAssociationMixin<OrderInstance>;
  getSwapDeal: BelongsToGetAssociationMixin<SwapDealInstance>;
  SwapDeal?: SwapDealInstance;
  makerOrder?: OrderInstance;
  takerOrder?: OrderInstance;
  createdAt: Date;
}

/* Node */
export type NodeCreationAttributes = NodeConnectionInfo;

export type NodeAttributes = NodeCreationAttributes & {
  id: number;
  banned: boolean;
  addressesText: string;
  lastAddressText: string;
  lastAddress: Address;
  consecutiveConnFailures: number;
};

export interface NodeInstance extends Model<NodeAttributes, NodeCreationAttributes>, NodeAttributes {
  reputationScore: number;
}

/* Pairs */
export type PairCreationAttributes = Pair;

export type PairAttributes = PairCreationAttributes & {
  id: string;
};

export interface PairInstance extends Model<PairAttributes, PairCreationAttributes>, PairAttributes {}

/* Reputation events */
export type ReputationEventCreationAttributes = {
  event: ReputationEvent;
  nodeId: number;
};

export type ReputationEventAttributes = ReputationEventCreationAttributes & {
  createdAt: number;
  id: number;
};

export interface ReputationEventInstance
  extends Model<ReputationEventAttributes, ReputationEventCreationAttributes>,
    ReputationEventAttributes {}

/* Passwords */
export type PasswordCreationAttributes = {
  encryptedPassword: string;
  currency?: string;
  swapClient: SwapClientType;
};

export type PasswordAttributes = PasswordCreationAttributes & {
  createdAt: number;
  id: number;
};

export interface PasswordInstance extends Model<PasswordAttributes, PasswordCreationAttributes>, PasswordAttributes {}
