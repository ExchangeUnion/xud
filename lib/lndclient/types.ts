/** The configurable options for the lnd client. */

export type LndClientConfig = {
  disable: boolean;
  certpath: string;
  macaroonpath: string;
  host: string;
  port: number;
  nomacaroons: boolean;
  cltvdelta: number;
};

/** General information about the state of this lnd client. */
export type LndInfo = {
  status: string;
  error?: string;
  channels?: ChannelCount;
  chains?: Chain[];
  blockheight?: number;
  uris?: string[];
  version?: string;
  alias?: string;
};

export type ChannelCount = {
  active: number;
  inactive?: number;
  pending: number;
  closed: number;
};

export type Chain = {
  network: string;
  chain: string;
};

export type ClientMethods =
  | 'close'
  | 'waitForReady'
  | 'makeClientStreamRequest'
  | 'makeBidiStreamRequest'
  | 'makeServerStreamRequest'
  | 'makeUnaryRequest';
