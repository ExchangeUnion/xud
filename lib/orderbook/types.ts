import { SwapClientType } from '../constants/enums';
import { SwapFailure, SwapSuccess } from '../swaps/types';

export type OrderBookThresholds = {
  minQuantity: number;
};

export type OrderMatch = {
  maker: Order;
  taker: OwnOrder;
};

export type MatchingResult = {
  matches: OrderMatch[];
  remainingOrder?: OwnOrder;
};

export type PlaceOrderResult = {
  internalMatches: OwnOrder[];
  swapSuccesses: SwapSuccess[];
  swapFailures: SwapFailure[];
  remainingOrder?: OwnOrder;
};

export type PlaceOrderEvent = {
  type: PlaceOrderEventType;
  order?: Order;
  swapSuccess?: SwapSuccess;
  swapFailure?: SwapFailure;
};

export enum PlaceOrderEventType {
  Match,
  SwapSuccess, // eslint-disable-line @typescript-eslint/no-shadow
  RemainingOrder,
  SwapFailure, // eslint-disable-line @typescript-eslint/no-shadow
}

/** An order without a price that is intended to match against any available orders on the opposite side of the book for its trading pair. */
type MarketOrder = {
  /** The quantity denominated in satoshis (or equivalent) for the order. */
  quantity: number;
  /** A trading pair symbol with the base currency first followed by a '/' separator and the quote currency */
  pairId: string;
  /** Whether the order is a buy (if `true`) or a sell (if `false`). */
  isBuy: boolean;
};

/** A limit order with a specified price that will enter the order book if it is not immediately matched. */
type LimitOrder = MarketOrder & {
  /** The price for the order expressed in units of the quote currency. */
  price: number;
};

/** Properties that can be used to uniquely identify and fetch an order within an order book. */
export type OrderIdentifier = Pick<MarketOrder, 'pairId'> & {
  /** The global identifier for this order on the XU network. */
  id: string;
};

/** Properties that apply only to orders placed by the local xud. */
type Local = {
  /** A local identifier for the order. */
  localId: string;
};

/** Properties that apply only to orders placed by remote peers. */
type Remote = {
  /** The nodePubKey of the node which created this order. */
  peerPubKey: string;
};

/** Properties that uniquely identify an order and make it ready to enter the order book. */
type Stamp = OrderIdentifier & {
  /** Epoch timestamp when this order was created locally. */
  createdAt: number;
  /** The number of satoshis (or equivalent) initially available for the order, before any actions such as trades reduced the available quantity. */
  initialQuantity: number;
};

export type SortableOrder = {
  price: number;
  createdAt: number;
};

export type OwnMarketOrder = MarketOrder & Local;

export type OwnLimitOrder = LimitOrder & Local;

/** A local order that may enter the order book. */
export type OwnOrder = OwnLimitOrder &
  Stamp & {
    /** The amount of an order that is on hold pending swap execution. */
    hold: number;
  };

/** A peer order that may enter the order book. */
export type PeerOrder = LimitOrder & Stamp & Remote;

export type Order = OwnOrder | PeerOrder;

/** An outgoing local order which only includes fields that are relevant to peers. */
export type OutgoingOrder = Pick<
  OwnOrder,
  Exclude<keyof OwnOrder, 'localId' | 'createdAt' | 'hold' | 'initialQuantity'>
> & {
  replaceOrderId?: string;
};

/** An incoming peer order which only includes fields that are relevant to us. */
export type IncomingOrder = OutgoingOrder & Remote;

/** A reference to a portion of an existing order. */
export type OrderPortion = OrderIdentifier & {
  quantity: number;
  localId?: string;
};

export type OrderInvalidation = OrderIdentifier & {
  /** The quantity of the order being removed, or the entire order if quantity is undefined */
  quantity?: number;
};

export type Currency = {
  /** The ticker symbol for this currency such as BTC, LTC, ETH, etc... */
  id: string;
  /** The payment channel network client to use for executing swaps. */
  swapClient: SwapClientType;
  /**
   * The number of places to the right of the decimal point of the smallest subunit of the currency, For example, BTC, LTC, and others
   * where the smallest subunits (satoshis) are 0.00000001 full units (bitcoins) have 8 decimal places. ETH has 18. This can be thought
   * of as the base 10 exponent of the smallest subunit expressed as a positive integer.
   */
  decimalPlaces: number;
  /** The contract address for layered tokens such as ERC20. */
  tokenAddress?: string;
};

export type Pair = {
  /* The base currency that is bought and sold for this trading pair. */
  baseCurrency: string;
  /* The currency used to quote a price for the base currency. */
  quoteCurrency: string;
};

export function isOwnOrder(order: Order): order is OwnOrder {
  return (order as PeerOrder).peerPubKey === undefined && typeof (order as OwnOrder).localId === 'string';
}

export function isPeerOrder(order: Order): order is PeerOrder {
  return (order as OwnOrder).localId === undefined && typeof (order as PeerOrder).peerPubKey === 'string';
}
