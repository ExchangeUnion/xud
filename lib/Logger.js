const fs = require('fs');
const path = require('path');
const winston = require('winston');
const util = require('./utils/utils');

class Logger {
  constructor({ level, logDir, context }) {
    this.level = level || Logger.defaultLevel;
    this.logDir = logDir || Logger.defaultLogDir;
    this.context = context || Logger.contexts.GLOBAL;

    if (!fs.existsSync(this.logDir)) {
      fs.mkdirSync(this.logDir);
    }
    this.logger = winston.createLogger({
      level: this.level,
      transports: [
        new (winston.transports.Console)({
          timestamp: util.getTsString,
          colorize: true,
        }),
        new (winston.transports.File)({
          filename: path.join(this.logDir, Logger.contextsFilename[this.context]),
          timestamp: util.getTsString,
        }),
      ],
    });

    this.error = this.error.bind(this);
    this.warn = this.warn.bind(this);
    this.info = this.info.bind(this);
    this.debug = this.debug.bind(this);
  }

  log(level, msg) {
    this.logger.log(level, msg);
  }

  error(msg) {
    this.log(Logger.levelsVal[Logger.levels.ERROR], msg);
  }

  warn(msg) {
    this.log(Logger.levelsVal[Logger.levels.WARN], msg);
  }

  info(msg) {
    this.log(Logger.levelsVal[Logger.levels.INFO], msg);
  }

  debug(msg) {
    this.log(Logger.levelsVal[Logger.levels.DEBUG], msg);
  }
}

Logger.levels = {
  ERROR: 0,
  WARN: 1,
  INFO: 2,
  DEBUG: 3,
};

Logger.levelsVal = {
  [Logger.levels.ERROR]: 'error',
  [Logger.levels.WARN]: 'warn',
  [Logger.levels.INFO]: 'info',
  [Logger.levels.DEBUG]: 'debug',
};

Logger.contexts = {
  GLOBAL: 0,
};

Logger.contextsFilename = {
  [Logger.contexts.GLOBAL]: 'xunion.log',
};

Logger.defaultLogDir = 'logs';

Logger.defaultLevel = process.env.NODE_ENV === 'production'
  ? Logger.levelsVal[Logger.levels.INFO]
  : Logger.levelsVal[Logger.levels.DEBUG];

Logger.global = new Logger({ context: Logger.contexts.GLOBAL });
module.exports = Logger;
