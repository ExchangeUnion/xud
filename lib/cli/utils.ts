import colors from 'colors/safe';
import { accessSync, watch } from 'fs';
import os from 'os';
import path from 'path';

const SATOSHIS_PER_COIN = 10 ** 8;

export function getDefaultCertPath() {
  switch (os.platform()) {
    case 'win32': {
      const homeDir = process.env.LOCALAPPDATA!;
      return path.join(homeDir, 'Xud', 'tls.cert');
    }
    case 'darwin': {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
    default: {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
  }
}

export function shorten(key: string, length = 10) {
  if (key.length <= length * 2 + 3) {
    return key;
  }
  return `${key.slice(0, length)}...${key.slice(key.length - length)}`;
}

export function trim(key: string, length = 10) {
  if (key.length <= length + 3) {
    return '';
  }
  return `${key.slice(0, length)}...`;
}

export const generateHeaders = (headers: string[]) => {
  return headers.map((header) => {
    return colors.blue(header);
  });
};

/** Returns a number of coins as an integer number of satoshis. */
export const coinsToSats = (coinsQuantity: number) => {
  return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
};

/** Returns a number of satoshis as a string representation of coins with up to 8 decimal places. */
export const satsToCoinsStr = (satsQuantity: number) => {
  return (satsQuantity / SATOSHIS_PER_COIN).toFixed(8).replace(/\.?0+$/, '');
};

/** Waits up to 5 seconds for the tls.cert file to be created in case this is the first time xud has been run. */
export const waitForCert = (certPath: string) => {
  return new Promise<void>((resolve, reject) => {
    try {
      accessSync(certPath);
      resolve();
    } catch (err) {
      if (err.code === 'ENOENT') {
        const certDir = path.dirname(certPath);
        const certFilename = path.basename(certPath);
        const fsWatcher = watch(certDir, (event, filename) => {
          if (event === 'change' && filename === certFilename) {
            clearTimeout(timeout);
            fsWatcher.close();
            resolve();
          }
        });
        const timeout = setTimeout(() => {
          fsWatcher.close();
          reject(`timed out waiting for cert to be created at ${certPath}`);
        }, 5000);
      } else {
        // we handle errors due to file not existing, otherwise reject
        reject(err);
      }
    }
  });
};

export function showSeed(seedMnemonicList: string[]) {
  const WORDS_PER_ROW = 4;
  const numberedMnemonic = seedMnemonicList.map((value, index) => {
    return `${index >= 9 ? '' : ' '}${index + 1}. ${value.padEnd(10)}`;
  });
  console.log('----------------------BEGIN XUD SEED---------------------');
  for (let n = 0; n < seedMnemonicList.length / WORDS_PER_ROW; n += 1) {
    console.log(numberedMnemonic.slice(n * WORDS_PER_ROW, (n + 1) * WORDS_PER_ROW).join(' '));
  }
  console.log('-----------------------END XUD SEED----------------------\n');

  console.log(`
Please write down your 24 word mnemonic. It will allow you to recover your xud \
node key and on-chain funds for the initialized wallets listed above should you \
forget your password or lose your device. Off-chain funds in channels can NOT \
be recovered with it and must be backed up and recovered separately. Keep it \
somewhere safe, it is your ONLY backup in case of data loss.
`);
}
