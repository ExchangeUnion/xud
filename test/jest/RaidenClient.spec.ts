import RaidenClient from '../../lib/raidenclient/RaidenClient';
import { RaidenClientConfig, TokenPaymentResponse } from '../../lib/raidenclient/types';
import Logger from '../../lib/Logger';
import { SwapDeal } from '../../lib/swaps/types';
import { UnitConverter } from '../../lib/utils/UnitConverter';
import { CurrencyInstance } from '../../lib/db/types';
import { getValidDeal } from '../utils';
import { PaymentState } from '../../lib/swaps/SwapClient';

const getValidTokenPaymentResponse = () => {
  return {
    amount: 100000000000000,
    initiator_address: '0x7ed0299Fa1ADA71D10536B866231D447cDFa48b9',
    secret_hash: '0xb8a0243672b503714822b454405de879e2b8300c7579d60295c35607ffd5613e',
    secret: '0x9f345e3751d8b7f38d34b7a3dd636a9d7a0c2f36d991615e6653501f30c6ec56',
    identifier: 3820989367401102,
    hashalgo: 'SHA256',
    target_address: '0x2B88992DEd5C96aa7Eaa9CFE1AE52350df7dc5DF',
    token_address: '0x4c354C76d5f73A63a90Be776897DC81Fb6238772',
  };
};

const channelBalance1 = 250000000000000000;
const channelBalance2 = 50000000000000000;
const channelBalanceTokenAddress = '0xEA674fdDe714fd979de3EdF0F56AA9716B898ec8';
const getChannelsResponse = [
  {
    token_network_identifier: '0xE5637F0103794C7e05469A9964E4563089a5E6f2',
    channel_identifier: 1,
    partner_address: '0x61C808D82A3Ac53231750daDc13c777b59310bD9',
    token_address: channelBalanceTokenAddress,
    balance: channelBalance1,
    total_deposit: 350000000000000000,
    state: 'opened',
    settle_timeout: 100,
    reveal_timeout: 30,
  },
  {
    token_network_identifier: '0xE5637F0103794C7e05469A9964E4563089a5E6f2',
    channel_identifier: 2,
    partner_address: '0x2A4722462bb06511b036F00C7EbF938B2377F446',
    token_address: channelBalanceTokenAddress,
    balance: channelBalance2,
    total_deposit: 350000000000000000,
    state: 'opened',
    settle_timeout: 100,
    reveal_timeout: 30,
  },
  {
    token_network_identifier: '0xE5637F0103794C7e05469A9964E4563089a5E6f2',
    channel_identifier: 3,
    partner_address: '0x3b1c3C1568C848b3C12c88e2aF5E5CAa0b62071A',
    token_address: channelBalanceTokenAddress,
    balance: 1000000,
    total_deposit: 35000000,
    state: 'closed',
    settle_timeout: 100,
    reveal_timeout: 30,
  },
];

const wethTokenAddress = '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2';
const currencyInstances = [
  { id: 'WETH', tokenAddress: wethTokenAddress },
];

jest.mock('../../lib/Logger');
describe('RaidenClient', () => {
  let raiden: RaidenClient;
  let config: RaidenClientConfig;
  let raidenLogger: Logger;
  let unitConverter: UnitConverter;

  beforeEach(() => {
    config = {
      disable: false,
      host: '127.0.0.1',
      port: 1234,
      keystorepath: '',
    };
    raidenLogger = new Logger({});
    raidenLogger.info = jest.fn();
    raidenLogger.error = jest.fn();
    raidenLogger.debug = jest.fn();
    unitConverter = new UnitConverter();
    unitConverter.init();
  });

  afterEach(async () => {
    jest.clearAllMocks();
    await raiden.close();
  });

  describe('sendPayment', () => {
    test('it removes 0x from secret', async () => {
      raiden = new RaidenClient({ unitConverter, config, directChannelChecks: true, logger: raidenLogger });
      await raiden.init(currencyInstances as CurrencyInstance[]);
      const validTokenPaymentResponse: TokenPaymentResponse = getValidTokenPaymentResponse();
      raiden['tokenPayment'] = jest.fn()
          .mockReturnValue(Promise.resolve(validTokenPaymentResponse));
      raiden.tokenAddresses.get = jest.fn().mockReturnValue(validTokenPaymentResponse.token_address);
      const deal: SwapDeal = getValidDeal();
      await expect(raiden.sendPayment(deal))
        .resolves.toMatchSnapshot();
    });

    test('it rejects in case of empty secret response', async () => {
      raiden = new RaidenClient({ unitConverter, config, directChannelChecks: true, logger: raidenLogger });
      await raiden.init(currencyInstances as CurrencyInstance[]);
      const invalidTokenPaymentResponse: TokenPaymentResponse = {
        ...getValidTokenPaymentResponse(),
        secret: '',
      };
      raiden['tokenPayment'] = jest.fn()
          .mockReturnValue(Promise.resolve(invalidTokenPaymentResponse));
      raiden.tokenAddresses.get = jest.fn().mockReturnValue(invalidTokenPaymentResponse.token_address);
      const deal: SwapDeal = getValidDeal();
      await expect(raiden.sendPayment(deal))
        .rejects.toMatchSnapshot();
    });
  });

  describe('openChannel', () => {
    let peerRaidenAddress: string;
    let units: number;
    let currency: string;

    beforeEach(() => {
      peerRaidenAddress = '0x10D8CCAD85C7dc123090B43aA1f98C00a303BFC5';
      units = 5000000;
      currency = 'WETH';
    });

    test('it fails when tokenAddress for currency not found', async () => {
      expect.assertions(1);
      raiden = new RaidenClient({ unitConverter, config, directChannelChecks: true, logger: raidenLogger });
      await raiden.init([] as CurrencyInstance[]);
      try {
        await raiden.openChannel({
          units,
          currency,
          peerIdentifier: peerRaidenAddress,
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    test('it throws when openChannel fails', async () => {
      expect.assertions(1);
      raiden = new RaidenClient({ unitConverter, config, directChannelChecks: true, logger: raidenLogger });
      const peerRaidenAddress = '0x10D8CCAD85C7dc123090B43aA1f98C00a303BFC5';
      const currency = 'WETH';
      const mockTokenAddresses = new Map<string, string>();
      mockTokenAddresses.set('WETH', wethTokenAddress);
      raiden.tokenAddresses = mockTokenAddresses;
      raiden['openChannelRequest'] = jest.fn().mockImplementation(() => {
        throw new Error('openChannelRequest error');
      });
      await raiden.init(currencyInstances as CurrencyInstance[]);
      try {
        await raiden.openChannel({
          units,
          currency,
          peerIdentifier: peerRaidenAddress,
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    test('it opens a channel', async () => {
      expect.assertions(2);
      raiden = new RaidenClient({ unitConverter, config, directChannelChecks: true, logger: raidenLogger });
      const peerRaidenAddress = '0x10D8CCAD85C7dc123090B43aA1f98C00a303BFC5';
      const currency = 'WETH';
      const mockTokenAddresses = new Map<string, string>();
      mockTokenAddresses.set('WETH', wethTokenAddress);
      raiden.tokenAddresses = mockTokenAddresses;
      raiden['openChannelRequest'] = jest.fn().mockReturnValue(Promise.resolve());
      await raiden.init(currencyInstances as CurrencyInstance[]);
      await raiden.openChannel({
        units,
        currency,
        peerIdentifier: peerRaidenAddress,
      });
      expect(raiden['openChannelRequest']).toHaveBeenCalledTimes(1);
      expect(raiden['openChannelRequest']).toHaveBeenCalledWith({
        partner_address: peerRaidenAddress,
        token_address: wethTokenAddress,
        total_deposit: units,
        settle_timeout: 500,
      });
    });
  });

  describe('lookupPayment', () => {
    const paymentHash = '63699fb42306ea693c7c9c038c18ecc8c7dbc8095b8ddd8d2e4421f2ce7b4c0c';

    test('it detects payment in pending state', async () => {
      raiden = new RaidenClient({ unitConverter, config, directChannelChecks: true, logger: raidenLogger });
      const peerRaidenAddress = '0x10D8CCAD85C7dc123090B43aA1f98C00a303BFC5';
      const currency = 'WETH';
      const mockTokenAddresses = new Map<string, string>();
      mockTokenAddresses.set('WETH', wethTokenAddress);
      raiden.tokenAddresses = mockTokenAddresses;
      raiden['getPendingTransfers'] = jest.fn().mockReturnValue(Promise.resolve([{
        payment_identifier: RaidenClient['getIdentifier'](paymentHash),
      }]));
      raiden['getPaymentEvents'] = jest.fn();
      await raiden.init(currencyInstances as CurrencyInstance[]);
      await expect(raiden.lookupPayment(paymentHash, currency, peerRaidenAddress))
        .resolves.toHaveProperty('state', PaymentState.Pending);
      expect(raiden['getPendingTransfers']).toHaveBeenCalledTimes(1);
      expect(raiden['getPendingTransfers']).toHaveBeenCalledWith(currency, peerRaidenAddress);
      expect(raiden['getPaymentEvents']).toHaveBeenCalledTimes(0);
    });

    test('it checks if payment has failed or completed if it is not pending', async () => {
      raiden = new RaidenClient({ unitConverter, config, directChannelChecks: true, logger: raidenLogger });
      const peerRaidenAddress = '0x10D8CCAD85C7dc123090B43aA1f98C00a303BFC5';
      const currency = 'WETH';
      const mockTokenAddresses = new Map<string, string>();
      mockTokenAddresses.set('WETH', wethTokenAddress);
      raiden.tokenAddresses = mockTokenAddresses;
      raiden['getPendingTransfers'] = jest.fn().mockReturnValue(Promise.resolve([]));
      raiden['getPaymentEvents'] = jest.fn().mockReturnValue(Promise.resolve([]));
      await raiden.init(currencyInstances as CurrencyInstance[]);
      await raiden.lookupPayment(paymentHash, currency, peerRaidenAddress);
      expect(raiden['getPendingTransfers']).toHaveBeenCalledTimes(1);
      expect(raiden['getPaymentEvents']).toHaveBeenCalledTimes(1);
      expect(raiden['getPaymentEvents']).toHaveBeenCalledWith(currency, peerRaidenAddress);
    });
  });

  test('channelBalance calculates the total balance of open channels for a currency', async () => {
    raiden = new RaidenClient({ unitConverter, config, directChannelChecks: true, logger: raidenLogger });
    await raiden.init(currencyInstances as CurrencyInstance[]);
    raiden.tokenAddresses.get = jest.fn().mockReturnValue(channelBalanceTokenAddress);
    raiden['getChannels'] = jest.fn()
        .mockReturnValue(Promise.resolve(getChannelsResponse));
    await expect(raiden.channelBalance('WETH')).resolves.toHaveProperty('balance', 30000000);
  });
});
