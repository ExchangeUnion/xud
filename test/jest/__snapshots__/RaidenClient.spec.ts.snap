// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaidenClient sendPayment rejects in case of empty secret response 1`] = `
Object {
  "code": "5.8",
  "message": "raiden TokenPaymentResponse is invalid",
}
`;

exports[`RaidenClient sendPayment removes 0x from secret 1`] = `"9f345e3751d8b7f38d34b7a3dd636a9d7a0c2f36d991615e6653501f30c6ec56"`;
